src/algebra/gcd_monoid.lean

```diff
@@ -632,7 +632,35 @@ by rw [gcd_comm _ a, gcd_comm _ a, gcd_eq_of_dvd_sub_right h]
 end integral_domain
 
 section constructors
-variables [comm_cancel_monoid_with_zero α] [nontrivial α] [normalization_monoid α]
+noncomputable theory
+
+variables [comm_cancel_monoid_with_zero α] [nontrivial α]
+
+/-- Define `normalization_monoid` on a structure from a `monoid_hom` inverse to `associates.mk`. -/
+def normalization_monoid_of_monoid_hom_right_inverse [decidable_eq α] (f : associates α →* α)
+  (hinv : function.right_inverse f associates.mk) :
+  normalization_monoid α :=
+{ norm_unit := λ a, if a = 0 then 1 else
+    classical.some (associates.mk_eq_mk_iff_associated.1 (hinv (associates.mk a)).symm),
+  norm_unit_zero := if_pos rfl,
+  norm_unit_mul := λ a b ha hb, by {
+    rw [if_neg (mul_ne_zero ha hb), if_neg ha, if_neg hb, units.ext_iff, units.coe_mul],
+    apply mul_left_cancel' (mul_ne_zero ha hb),
+    conv_rhs { rw [← mul_assoc, mul_comm, mul_assoc, mul_comm b, mul_comm, ← mul_assoc], },
+    rw [classical.some_spec (associates.mk_eq_mk_iff_associated.1 (hinv (associates.mk a)).symm),
+      classical.some_spec (associates.mk_eq_mk_iff_associated.1 (hinv (associates.mk (a * b))).symm),
+      mul_assoc,
+      classical.some_spec (associates.mk_eq_mk_iff_associated.1 (hinv (associates.mk b)).symm),
+      ← monoid_hom.map_mul, associates.mk_mul_mk] },
```

Comment:
With the `suffices`, the block of `rw [mul_assoc, mul_comm]` is a bit less mysterious. I'm not happy how we have to wrap the lines, maybe an extra abbreviation is in order for `classical.some{_spec} associated_map_mk`.
```suggestion
    suffices : (a * b) * ↑(classical.some (associated_map_mk hinv (a * b))) =
      (a * ↑(classical.some (associated_map_mk hinv a))) *
      (b * ↑(classical.some (associated_map_mk hinv b))),
    { apply mul_left_cancel' (mul_ne_zero ha hb) _,
      simpa only [mul_assoc, mul_comm, mul_left_comm] using this },
    rw [classical.some_spec (associated_map_mk hinv a),
      classical.some_spec (associated_map_mk hinv (a * b)),
      classical.some_spec (associated_map_mk hinv b),
      ← monoid_hom.map_mul, associates.mk_mul_mk] },
```